File: day13/day13.kt
import java.io.File

fun main() {
    val lines = File("day13.txt").readLines()
    val N = lines[0].toInt()
    val buses = lines[1].split(',').map { if (it != "x") it.toInt() else null }

    val wait = buses.filterNotNull().map { it - N % it to it }.toMap()
    val key = wait.keys.minOrNull()!!
    println(key * wait.get(key)!!)

    val lookup = buses.mapIndexed { k, v ->
        if (v != null) v to k + 1 else null }.filterNotNull()
    var start = 0L; var step = 1L
    for (idx in 1..lookup.size) {
        while (!lookup.take(idx).all { (k, v) ->
            k - start % k == (v % k).toLong() }) start += step
        step *= lookup[idx - 1].first
    }
    println(start + 1)
}

File: 2020/day13/day13.py
lines = list(open('day13.txt'))
N = int(lines[0])
buses = [int(nr) if nr != 'x' else None for nr in lines[1].split(',')]

wait = {nr - N % nr: nr for nr in buses if nr is not None}
key = min(wait)
print(key * wait[key])

lookup = [(v, k + 1) for k, v in enumerate(buses) if v is not None]
verify = lambda a, n: all(k - n % k == v % k for k, v in a)
start, step = 0, 1
for idx in range(len(lookup)):
    while not verify(lookup[:idx+1], start):
        start += step
    step *= lookup[idx][0]
print(start + 1)
